[AIRFLOW-5172] Add choice of interval edge scheduling

iroddis opened the pull request:
Make sure you have checked _all_ steps below.

### Jira

- [X] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-5172

### Description

- [X] Here are some details about my PR, including screenshots of any UI changes:

This change introduces the attribute `schedule_interval_edge`, a string containing either 'start' or 'end', to DAGs. The scheduler uses the value to determining if a DAG should be scheduled at the start or the end of the schedule interval.

A parameter with the same name was also added to the default_airflow.cfg in the [scheduler] section. 

### Tests

- [X] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

tests.jobs.test_scheduler_job:SchedulerJobTest.test_schedule_interval_edge

### Commits

- [X] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [X] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### Code Quality

- [X] Passes `flake8`

ashb commented on docs/scheduler.rst:
```diff
@@ -94,6 +94,75 @@ should be triggered and come to a crawl. It might also create undesired
+A few examples:
+
+- A DAG with ``schedule_interval='@hourly'``: The DAG run that processes
+  2019-08-16 17:00 will start running at 2019-08-16 18:00.
```
```suggestion
  2019-08-16 17:00 will start running just after 2019-08-16 17:59:59 -- i.e. once that hour is over.
```

ashb commented on docs/scheduler.rst:
```diff
@@ -94,6 +94,75 @@ should be triggered and come to a crawl. It might also create undesired
+    # Processing: 2019-08-08
+    # Execution: Shortly after 2019-08-08 00:00
+    # Will execute at the start of the day to be processed
+    sod = DAG(
```
This is an unfortunate abbreviation in British English :)

```suggestion
    start_of_day = DAG(
```

ashb commented on airflow/config_templates/default_airflow.cfg:
```diff
@@ -511,6 +511,10 @@ scheduler_zombie_task_threshold = 300
 # DAG definition (catchup)
 catchup_by_default = True
 
+# Should DAGs be schedule to execute at the 'start' or 'end'
```
```suggestion
# Should DAGs be scheduled to execute at the 'start' or 'end'
```

ashb commented on airflow/config_templates/default_airflow.cfg:
```diff
@@ -511,6 +511,10 @@ scheduler_zombie_task_threshold = 300
 
+# Should DAGs be schedule to execute at the 'start' or 'end'
+# of their schedule interval.
+execute_at_interval_end = True
```
I wonder if `schedule_at_interval_end` is a better/clearer name.

I'm probably splitting hairs

iroddis replied:
I'm all for hair splitting if it means clearer usage down the road, and is more consistent with established nomenclature.

Part of my confusion is that `schedule` seams to have multiple meanings: A job scheduled daily will be scheduled to run once the day has finished. So there's the declared schedule, and then the implied schedule (execution time).

I have no vested interest in the name, just let me know if you'd like to change it.

Either way, maybe it would be worth adding (in a separate PR) a dictionary of terms and precise definitions? I'd be happy submit that after this.

ashb replied:
All these terms are confusing and overlapping.

The reason I wanted to avoid `execute` is because the `execution_date` property of the task  will remain constant non matter what this setting. So I think this does need to change, but I'm not at all convinced my wording is better.

> Either way, maybe it would be worth adding (in a separate PR) a dictionary of terms and precise definitions? I'd be happy submit that after this.

Yes, that would be great!

iroddis replied:
I've updated the name, I like it, and it makes sense. Also made a few more very small documentation changes, and squashed all the commits. Please let me know if you'd like to change anything else. 

Do you think there's any way this could get cherry picked for the next stable release? Right now I'm using a hacked up local version to get this behaviour, and it would be great to get back to upstream stable.

ashb reviewed (approved):
Happy with the code on this change, I'm opening a thread on the mailing list to discuss this.

(Committers: please don't merge without checking that thread first)

feng-tao commented on airflow/config_templates/default_airflow.cfg:
```diff
@@ -511,6 +511,10 @@ scheduler_zombie_task_threshold = 300
 
+# Should DAGs be scheduled to execute at the 'start' or 'end'
+# of their schedule interval.
+schedule_at_interval_end = True
```
+1 on the default to be false as this is the intuitive / common understanding on how Airflow scheduling works.

iroddis replied:
The PR is designed to have zero impact to current behaviours, to avoid surprises for devops everywhere and to increase the chance that the PR could be accepted for the next 1.x.x stable release.

Would it make more sense to leave this PR as is, and then submit a second PR to make the default False at a later date, for 2.0?

JCoder01 commented on airflow/jobs/scheduler_job.py:
```diff
@@ -649,7 +649,10 @@ def create_dag_run(self, dag, session=None):
+                if dag.schedule_at_interval_end:
+                    period_end = dag.following_schedule(next_run_date)
+                else:  # Schedule it at the start of the interval
+                    period_end = next_run_date
```
Does similar logic need to be applied to manage_slas?

Commit:
[AIRFLOW-5172] Add core schedule_at_interval_end code

Adds a boolean DAG attribute schedule_at_interval_end. If true
(default), jobs will be scheduled to run shortly after the
schedule_interval has elapsed (current airflow behaviour).

If False, the DAG will be scheduled to run at the beginning
of the schedule interval (new behaviour). This allows for
scheduling that more closely mirrors a typical cron schedule.

Adds a period_end method to DAG to return the datetime of
the end of a period.

Adds support in the SchedulerJob for this feature.

Adds a default of True to the default config in the [scheduler]
section.

Add unit test for start/end interval scheduling.

Updates docs/scheduler.rst to explain the new behaviour.

Co-Authored-By: Ash Berlin-Taylor <ash_github@firemirror.com>
Update docs/scheduler.rst

Replacing straight following_schedule calls with period_end calls
as appropriate to handle the new start/end of schedule interval
scheduling.

dstandish commented on airflow/jobs/scheduler_job.py:
```diff
@@ -648,7 +648,7 @@ def create_dag_run(self, dag, session=None):
                 period_end = next_run_date
             elif next_run_date:
-                period_end = dag.following_schedule(next_run_date)
+                period_end = dag.period_end(next_run_date)
```
it seems there is an inconsistency in the language here.  

The function is called `period_end`, yet sometimes it returns the start of the interval. 

The name `period_end` for this variable reflects an assumption that dags are scheduled after the _end_ of the interval.  

That's why the code checks that `period_end <= timezone.utcnow()`: end of interval is in the past.

But language of this PR is in conflict with that.  

The parameter `schedule_at_interval_end` implies that the _interval_ doesn't change, but _where we schedule_ does.  So, we may schedule at start or end of "the interval".  But as written, if `schedule_at_interval_end=False`, it will in general be the case that `period_end==excecution_date`, which implies that exec date is the end of "the interval" and not the start, and this is a contradiction.

It seems that what `period_end` represents in this code is more like `run_after_dttm` -- the datetime before which the dag may not be scheduled.  When `schedule_at_interval_end` is True, we can run after exec date + 1 interval; otherwise, we can run after exec date.  

So in this bit of code, we probably don't even need `period_end()` because we could do this:
```
            run_after_dttm = None
            if dag.schedule_interval == '@once':
                run_after_dttm = next_run_date
            elif next_run_date and not self.schedule_at_interval_end:
                run_after_dttm = next_run_date
            elif next_run_date and self.schedule_at_interval_end:
                run_after_dttm = dag.following_schedule(next_run_date)
```
And this:
```
            if next_run_date and run_after_dttm and run_after_dttm <= timezone.utcnow():
```

But elsewhere, it seems that `period_end()` function is used to mean `min_run_date` or `target_run_date` or `run_after_date`.  Perhaps a name like this would be clearer.

iroddis replied:
Nomenclature has absolutely been a pain in this change. The entire vocabulary around scheduling seems to have exploded, and I can't find a clear definition for all the terms. This is well outside the scope of this change, but having well defined terms for things like:

period_start_ts - Timestamp of the start of the period to be processed
period_end_ts - Timestamp of the end of the period to be processed
scheduled_execution_ts - Timestamp of when the execution should start
execution_ts - Timestamp of actual execution start

Many of the current names, like start_date, are datetimes, not dates. It's also not immediately obvious what "start" means: is it period_start_ts, or execution_ts?

I'm happy to change the functions to be whatever is popular, but I think there'd be real value in defining, clearly, and in a single spot, all of the timestamps around a particular DAG's execution, giving them meaningful names, and consolidating on that.

dstandish replied:
Yeah i'm not saying you are responsible for all confusing naming, just suggesting not to add to it ;)

`period_end` is a new function. 

And I think it is definitely in scope to change the local `period_end` variables since you are, in effect, changing their meaning in this PR.
